### PR TITLE
Rework shell and console connections

### PR DESCRIPTION
**Pull request by tsoutsman**
This PR is the first step in reworking the shell, adding a `shell_2` crate that is used by `console` for serial connections. It works and can be connected to using `screen` or a similar application. A couple of problems with the shell:

- [ ] I've added history deduplication to `noline` but haven't merged it into upstream yet. There's currently a patch in `Cargo.toml`.
- [ ] Autocompletion will need to be implemented upstream.
- [ ] Theseus handles Ctrl+C in a non-standard way and `noline` doesn't clear the current line when it encounters a Ctrl+C.
- [ ] If multiple keys are pressed in quick succession after connecting to the serial port, `noline` will fail. `noline` queries the terminal emulator for screen size using ANSII escape codes. However, the key press will be interspersed with the response from the terminal if the `port_to_tty_loop` isn't scheduled between the `noline` request and user key press. This isn't a major issue but I should probably document it somewhere.

I've started on a terminal emulator that uses `shell_2` but decided to first submit this PR as it's already massive as it is.

The PR is dependent on #657.

I've also done an ad hoc rework of `app_io` which I'm not a particular fan of. It leads to a number of `Arc<Arc<T>>` sandwiches. I also need to port over `immediate_input_echo`, `input_echo`, and `keyboard_echo` to use ANSII escape codes and the line discipline.

The PR also removes the `print` and `terminal_print` crates which were both legacy interfaces. They've been replaced with `app_io`. Most of the file changes are due to this.

### CODE REVIEW

File: Cargo.lock
Comment (kevinaboos) [inline Cargo.lock]: is there a strong need for this hashmap version? if not, i'd prefer to keep all the versions consistent (we can upgrade later) at v0.11.*

File: Cargo.toml
Comment (kevinaboos) [inline Cargo.toml]: please move this to a `theseus-os`-owned repo and change this patch to use that

File: applications/immediate_input_echo/src/lib.rs
Comment (kevinaboos) [inline applications/immediate_input_echo/src/lib.rs]: do you intend to fix this for this PR, or just remove the entire app?

File: applications/shell/src/lib.rs
Comment (kevinaboos) [inline applications/shell/src/lib.rs:638]: quick question: this appears to be for handling line-buffered stdin vs unbuffered stdin; is that now handled elsewhere? If not, we should leave a todo or some other kind of comment (or file an issue as a reminder)

File: applications/shell_2/src/internal.rs
Comment (kevinaboos) [inline applications/shell_2/src/internal.rs]: i think the typical term for these are "shell builtins", so we ought to use that term for this stuff instead of `internal`
https://en.wikipedia.org/wiki/Shell_builtin

File: applications/shell_2/src/error.rs
Comment (kevinaboos) [inline applications/shell_2/src/error.rs]: add docs describing each enum variant, especially the `Command` one and what the internal `isize` represents

File: applications/shell_2/src/internal.rs
Comment (kevinaboos) [inline applications/shell_2/src/internal.rs]: for all of these commands, instead of panicking, let's either:
1. print a message (and return an error) explaining that this builtin isn't yet implemented, we can optionally print a link to the Theseus github issue tracker with the suggestion to file an issue asking for it to be supported
2. print a message (and return an error) explaining that this builtin is not relevant or applicable to Theseus and will never be supported

File: applications/shell_2/src/lib.rs
Comment (kevinaboos) [inline applications/shell_2/src/lib.rs]: ? clarify please

File: applications/shell_2/src/lib.rs
Comment (kevinaboos) [inline applications/shell_2/src/lib.rs]: is there a use case where creating a new `Shell` and `run`ning it need to be separate actions?

File: applications/shell_2/src/lib.rs
Comment (kevinaboos) [inline applications/shell_2/src/lib.rs]: Rust typically uses a leading underscore to indicate something that isn't used. If that's not appropriate for this, i suggest something like `run_inner` or the like

File: applications/shell_2/src/lib.rs
Comment (kevinaboos) [inline applications/shell_2/src/lib.rs]: for what, setting the line discipline? Yea, i think that'd be a nice feature.

File: applications/shell_2/src/lib.rs
Comment (kevinaboos) [inline applications/shell_2/src/lib.rs]: same comment as above about the leading underscore. When looking at just this function in isolation, I can't tell if this is supposed to be an alternative to "commenting out" this method, or if it's an internal fn.

File: applications/shell_2/src/lib.rs
Comment (kevinaboos) [inline applications/shell_2/src/lib.rs]: bump; address these before merge or delay for later?

File: applications/shell_2/src/wrapper.rs
Comment (kevinaboos) [inline applications/shell_2/src/wrapper.rs]: this is quite strange with the assert statements. Can you shed some light on this?

File: kernel/app_io/Cargo.toml
Comment (kevinaboos) [inline kernel/app_io/Cargo.toml]: use v0.11 for consistency

File: kernel/async_channel/src/lib.rs
Comment (kevinaboos) [inline kernel/async_channel/src/lib.rs]: what's with these changes to the async_channel; are they necessary for the new shell/console design?

Note that there's also the `mpmc` queue or crossbeam's various `mpsc` queue types that might be better for your needs, depending on what features you actually need.

If the channel needs to change, i'd prefer to separate that out into a distinct PR so it's easier to track why things changed, rather than hiding it within a huge PR like this

File: kernel/async_channel/src/lib.rs
Comment (kevinaboos) [inline kernel/async_channel/src/lib.rs]: same question as above

File: kernel/console/Cargo.toml
Comment (kevinaboos) [inline kernel/console/Cargo.toml]: should probably add yourself as an author

File: kernel/console/src/lib.rs
Comment (kevinaboos) [inline kernel/console/src/lib.rs]: indeed

File: kernel/console/src/lib.rs
Comment (kevinaboos) [inline kernel/console/src/lib.rs]: here? it's already using shell_2 below. Or is this a general todo for the future? If so, we can file an issue to track that

File: kernel/console/src/lib.rs
Comment (kevinaboos) [inline kernel/console/src/lib.rs:140]: Currently the `kill` functionality doesn't work properly (it doesn't unwind the task and clean up its resources), so we definitely should not use it here. Instead, use some kind of shared state or message to inform those two tasks that they should exit cleanly, just as you would do with classic userspace pthreads.

File: kernel/console/src/lib.rs
Comment (kevinaboos) [inline kernel/console/src/lib.rs:151]: so, we've reached here when the remote client endpoint quits? (e.g., when i close the terminal emulator on the host side)
At that point, yes, we should clean up everything and return this to its initial state, in preparation for a brand new connection

File: kernel/console/src/lib.rs
Comment (kevinaboos) [inline kernel/console/src/lib.rs]: i know we haven't done a great job of this in the past, but are these two tasks blocking? It's pretty heavyweight to spawn so many tasks to handle each console connection, so i want to make sure they're not Runnable and endlessly spinning while waiting for new data

Comment (kevinaboos) [reply kernel/console/src/lib.rs]: (looks like they're using the async_channel, but i just want to be sure)

File: kernel/task/src/lib.rs
Comment (kevinaboos) [inline kernel/task/src/lib.rs]: I'm hesitant to introduce new runstates unnecessarily, as it adds lots of complexity to all the places where task runstates are used and checked. Is this absolutely mandatory, and if so, why?  I understand it's slightly semantically different than `Blocked`, but is it a true problem to reuse `Blocked` for this purpose?

File: kernel/scheduler/src/lib.rs
Comment (kevinaboos) [inline kernel/scheduler/src/lib.rs]: undo changes to this file before merge

File: kernel/tty/Cargo.toml
Comment (kevinaboos) [inline kernel/tty/Cargo.toml:6]: add author & description

File: kernel/tty/src/channel.rs
Comment (kevinaboos) [inline kernel/tty/src/channel.rs]: technically true, but they're separated in order to limit the capabilities of the caller, so let's not overload that

File: kernel/tty/src/channel.rs
Comment (kevinaboos) [inline kernel/tty/src/channel.rs:5]: this whole module isn't really tty-specific, right?. It looks like you want a channel effectively does what `read_to_end` does? 

If so, let's separate out all of these channel-specific changes into a different PR so we can analyze them in a standalone manner

File: kernel/tty/src/discipline.rs
Comment (kevinaboos) [inline kernel/tty/src/discipline.rs:21]: Q: does this render properly in rustdoc? i haven't seen this syntax before

File: kernel/tty/src/discipline.rs
Comment (kevinaboos) [inline kernel/tty/src/discipline.rs]: nit: maybe a bit better to use `Vec::clear()` ?

File: kernel/tty/src/discipline.rs
Comment (kevinaboos) [inline kernel/tty/src/discipline.rs]: the serial port impl already attempts to read multiple bytes (more specifically, `u16` input events) on every interrupt and send larger batches (up to 256 `u16`s for a total max of 512 bytes) per msg, so you could take a similar batching approach here

File: kernel/tty/src/lib.rs
Comment (kevinaboos) [inline kernel/tty/src/lib.rs:177]: i see what's happening, but i guess i'm confused by the usage of the term "immutable" here. hmm....

File: kernel/wait_queue/src/lib.rs
Comment (kevinaboos) [inline kernel/wait_queue/src/lib.rs]: 657 has been merged in now :tada:

File: theseus_features/Cargo.toml
Comment (kevinaboos) [inline theseus_features/Cargo.toml]: add a separate feature here, this removes the existing shell as an optional component.

File: kernel/app_io/src/lib.rs
Comment (kevinaboos) [inline kernel/app_io/src/lib.rs]: "full control to queues"? small typo?

**Review by kevinaboos — CHANGES_REQUESTED** [review]
Top-level comments:
* For specific test apps like `immediate_input_echo`, `keyboard_echo`, `less`, and `input_echo`, do you plan to fix them as part of this PR, remove them entirely, or re-implement them later?
* Add crate-level or module-level docs to the new files, e.g., `shell_2` and its modules
* There are _quite_ a few `unwrap` operations in the `shell_2`. Normally I don't mind being lazy and introducing panicking instructions in application code, but in this case the `shell` is kind of the fundamental interactive application since Theseus lacks a true desktop environment. So if the shell panics, it's quite hard to display something indicating that to the user. 
* [optional] rename `shell_2` to something more descriptive? "ng" or something, or perhaps a new name entirely. I had some ideas related to the "Ship of Theseus" theme, e.g., "hull" -- it's the _shell_ of a ship... lol
* `ImmutableWrite`/`Read` sound a bit like my concept of `LockableIo` -- can we combine those concepts or perhaps adapt `LockableIo` for your needs?

* Last but not least, how does this changeset interact with the "legacy" feature that allows kernel crates (without an app context) to also print to the "default" terminal? I'm primarily concerned about printing exceptions and other important messages both before and after the shell registers app_io streams. Without a desktop environment, there's no other way to do that, considering that the logger isn't available on most real hardware.

File: kernel/tty/src/discipline.rs
Comment (tsoutsman) [reply kernel/tty/src/discipline.rs:21]: A: Yes.
image

They are called [reference-style links](https://www.markdownguide.org/basic-syntax/#reference-style-links) and are [commonly used](https://github.com/rust-lang/rust/blob/e221616639fb87de9dca21e252ee8a2565ec51d0/library/std/src/thread/mod.rs#L136-L148) in `std`.

File: kernel/app_io/src/lib.rs
Comment (tsoutsman) [reply kernel/app_io/src/lib.rs]: I didn't write that but I updated it to be correct.

File: kernel/console/src/lib.rs
Comment (tsoutsman) [reply kernel/console/src/lib.rs]: Yes, the call to `read`/`receive` eventually boils down to `WaitQueue::wait_until`. We need separate tasks because there currently isn't a way to wait on two things at once.

File: kernel/console/src/lib.rs
Comment (tsoutsman) [reply kernel/console/src/lib.rs:151]: This was more of an "I'm not sure if dropping `port` is enough to return it to its initial state".

File: kernel/console/src/lib.rs
Comment (tsoutsman) [reply kernel/console/src/lib.rs:140]: I'm not sure how we should do this since they're blocked on a channel.

1. Spin in a loop using `try_receive` and checking an `AtomicBool`. This is terrible as it means `tty_to_port_loop` and `port_to_tty_loop` are always `Runnable`.
2. 
```rust
let mut data = [0; 256];
loop {
    let len = master.read(&mut data).unwrap();
    if should_exit.load(Ordering::SeqCst) {
        break;
    }
    port.lock().write(&data[..len]).unwrap();
}
```
and then write something to the buffer from `shell_loop` after setting `should_exit` to `true` to ensure it is tested. I'm not sure how we would do this for `port_to_tty_loop` as it's waiting on a serial port. Could we expose a function from `serial_port` that returns a write handle so that we can imitate input from the serial port?

File: applications/shell/src/lib.rs
Comment (tsoutsman) [reply applications/shell/src/lib.rs:638]: yep that's now handled by the canonical flag on the line discipline.

File: applications/shell_2/src/lib.rs
Comment (tsoutsman) [reply applications/shell_2/src/lib.rs]: I think this can be delayed till later. The double `Arc` is unfortunate but not too important. I haven't set the environment because currently, it behaves in a weird way where the child can modify the parent's environment. I think this should be changed but left it for another PR.

File: applications/shell_2/src/wrapper.rs
Comment (tsoutsman) [reply applications/shell_2/src/wrapper.rs]: Since `stdout.read` and `stdout.write` block, they should never return `Ok(0)`.

File: kernel/async_channel/src/lib.rs
Comment (tsoutsman) [reply kernel/async_channel/src/lib.rs]: They are because we use `async_channel` as an `mpmc`. `async_channel` will stop working as soon as a single sender or receiver is dropped which seems wrong.

File: kernel/tty/src/channel.rs
Comment (tsoutsman) [reply kernel/tty/src/channel.rs:5]: Not quite. This makes `read` blocking when operating on buffers. It will block on the first byte, and then use `try_read`/`try_write` to operate on the remainder of the bytes.

File: kernel/task/src/lib.rs
Comment (tsoutsman) [reply kernel/task/src/lib.rs]: If we press Ctrl+Z (suspend) while a task is blocked on something, it will be unblocked by the thing it is waiting for unless we have a separate run state.

FWIW (nothing), [Linux has a separate STOPPED state](https://www.cbtnuggets.com/blog/certifications/open-source/what-are-the-5-linux-process-states) (equivalent to suspended).

Comment (tsoutsman) [reply kernel/task/src/lib.rs]: There may be other reasons Linux has a separate state but I can't think of any.

**Comment by tsoutsman** [comment]
@kevinaboos 
> For specific test apps like immediate_input_echo, keyboard_echo, less, and input_echo, do you plan to fix them as part of this PR, remove them entirely, or re-implement them later?

I don't think it's possible to reimplement keyboard_echo. I can reimplement immediate_input_echo and input_echo however, this means they won't work with the current shell and GUI, only with `hull` over serial. Once we have a GUI terminal emulator that runs `hull` this won't be an issue. less can be implemented, but it will require a fair bit of work so I'd rather leave that for another PR.

> Add crate-level or module-level docs to the new files, e.g., shell_2 and its modules

:+1: I've added some basic docs. I'm not sure what else to add.

> There are quite a few unwrap operations in the shell_2. Normally I don't mind being lazy and introducing panicking instructions in application code, but in this case the shell is kind of the fundamental interactive application since Theseus lacks a true desktop environment. So if the shell panics, it's quite hard to display something indicating that to the user.

Should be done. There are still a couple of unwraps/excepts but they are for some pretty basic stuff like having `stdin` and `stdout`.

> rename shell_2 to something more descriptive? "ng" or something, or perhaps a new name entirely. I had some ideas related to the "Ship of Theseus" theme, e.g., "hull" -- it's the shell of a ship... lol

I like it. Goes in line with `captain`.

> ImmutableWrite/Read sound a bit like my concept of LockableIo -- can we combine those concepts or perhaps adapt LockableIo for your needs?

I'm not sure. For example `Slave` implements `ImmutableWrite/Read`, but internally it uses a queue not locks.

> Last but not least, how does this changeset interact with the "legacy" feature that allows kernel crates (without an app context) to also print to the "default" terminal? I'm primarily concerned about printing exceptions and other important messages both before and after the shell registers app_io streams. Without a desktop environment, there's no other way to do that, considering that the logger isn't available on most real hardware.

I'll add that functionality to `app_io`.

Aside from that, most of the "changes requested" are complete. I've left comments on the ones that haven't yet been resolved.

File: applications/shell_2/src/lib.rs
Comment (kevinaboos) [reply applications/shell_2/src/lib.rs]: right, the environment is shared with that of the "parent" task. If you don't want that behavior, you can clone the environment, set the child task's env to that new cloned instance, and then modify it.

File: applications/shell_2/src/wrapper.rs
Comment (kevinaboos) [reply applications/shell_2/src/wrapper.rs]: I see. Returning an error (e.g., "short write") might be a bit clearer than a panicking assert without a message.

File: kernel/async_channel/src/lib.rs
Comment (kevinaboos) [reply kernel/async_channel/src/lib.rs]: ah i see, thanks for identifying that. Looks like that was added in #277 (during a paper deadline crunch) so the author was likely only testing it with a single endpoint of each variety. Should be a simple enough fix: only mark the channel as disconnected when the _last_ Sender or Receiver gets dropped, not just any sender or receiver.

Comment (kevinaboos) [reply kernel/async_channel/src/lib.rs]: I filed #703 about this. For now you can leave this change in there as part of this PR.

File: kernel/console/src/lib.rs
Comment (kevinaboos) [reply kernel/console/src/lib.rs:151]: hmm, then what happens currently if you close the PTY on the linux host side and then try to re-open it?

File: kernel/task/src/lib.rs
Comment (kevinaboos) [reply kernel/task/src/lib.rs]: I see. So we either need a separate runstate or a "blocked semaphore" that counts how many times a task has been blocked (kind of like how we do preemption holds). 

Q: even with a separate runstate, consider how the following scenario plays out:
1. task is running as normal
2. task blocks itself upon something, e.g., waiting for I/O. 
    --> task is now Blocked
3. user suspends the task using the shell
    --> task is now Suspended (?)
4. the I/O operation completes, unblocking the task
   --> now what should happen? 

The task is no longer blocked but it's also not runnable... we'd need a sort of "stack" to track prior runstate transitions to accomplish this.

Alternatively, should "Suspended" be a completely separate bit of information rather than a runstate? That's potentially even more annoying to handle, but it's also more correct.

File: kernel/tty/src/channel.rs
Comment (kevinaboos) [reply kernel/tty/src/channel.rs:5]: oh right, i skimmed it too quickly. 

Could be useful on its own too, right? Perhaps consider moving it into a separate crate since it's essentially just a wrapper around a channel.

File: kernel/tty/src/discipline.rs
Comment (kevinaboos) [reply kernel/tty/src/discipline.rs:21]: neat, TIL you can combine those two forms of hyperlinking

File: kernel/app_io/src/lib.rs
Comment (kevinaboos) [reply kernel/app_io/src/lib.rs]: oh my apologies. Thanks for fixing it anyway.

**Comment by kevinaboos** [comment]
> I don't think it's possible to reimplement keyboard_echo. I can reimplement immediate_input_echo and input_echo however, this means they won't work with the current shell and GUI, only with `hull` over serial. Once we have a GUI terminal emulator that runs `hull` this won't be an issue. less can be implemented, but it will require a fair bit of work so I'd rather leave that for another PR.

Totally fine with me. Instead of commenting out their content, we can just move them to the `old_crates/` directory (and update. Alternatively, you could create an `app_io_ng` or whatever to service your new shell, such that the original `app_io` would still support those old applications. But I don't personally care about those old apps, only `less` is actually useful; it'd be easy to re-do it later if needed.

> > ImmutableWrite/Read sound a bit like my concept of LockableIo -- can we combine those concepts or perhaps adapt LockableIo for your needs?
> 
> I'm not sure. For example `Slave` implements `ImmutableWrite/Read`, but internally it uses a queue not locks.

np, just ignore it for now.

> Aside from that, most of the "changes requested" are complete. I've left comments on the ones that haven't yet been resolved.

Cool. I've left a few responses to comments as well. Let me know when things are ready to review again. 👍

File: kernel/console/src/lib.rs
Comment (tsoutsman) [reply kernel/console/src/lib.rs:151]: the host's terminal hangs

Comment (kevinaboos) [reply kernel/console/src/lib.rs:151]: ok yeah, so then we need to do a more thorough cleanup/reset procedure to allow another host terminal to connect.

File: applications/shell_2/src/lib.rs
Comment (tsoutsman) [reply applications/shell_2/src/lib.rs]: Turns out this is already done in `Task::new`. The old shell was intentionally overwriting the child's environment with a reference to the parent's environment.

File: kernel/task/src/lib.rs
Comment (tsoutsman) [reply kernel/task/src/lib.rs]: I added a separate `suspended` field to `Task` in 2f0d95c.

File: kernel/tty/src/channel.rs
Comment (tsoutsman) [reply kernel/tty/src/channel.rs:5]: Moved the methods to `async_channel` in 595948c. Now the module is just to convert the `async_channel::Error` to `core2::io::Error`.

File: kernel/console/src/lib.rs
Comment (tsoutsman) [reply kernel/console/src/lib.rs:151]: `console` doesn't detect the second connection, and since this isn't a pressing issue I'd rather leave it for a future PR.

**Comment by kevinaboos** [comment]
@tsoutsman Now that #707 is merged in and #703 is closed, hopefully the issues with async_channel here are resolved. (no rush)

**Comment by tsoutsman** [comment]
> @tsoutsman Now that https://github.com/theseus-os/Theseus/pull/707 is merged in and https://github.com/theseus-os/Theseus/issues/703 is closed, hopefully the issues with async_channel here are resolved. (no rush)

Done.

**Review by kevinaboos — APPROVED** [review]
Quite a few to-do items left, but we can go ahead and merge this in order to move forward and not let such a large changeset linger and get stale.

**Pull request merged**
